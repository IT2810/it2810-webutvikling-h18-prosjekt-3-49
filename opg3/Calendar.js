import React, { Component } from 'react';
import { StyleSheet, Text, View, FlatList, Button, TextInput} from 'react-native';
import { Agenda } from 'react-native-calendars';
import Storage from './Storage.js'

export default class Calendar extends Component {
  constructor(props) {
    super(props);

    let time = new Date();
    time = time.toISOString().split('T')[0];

    this.state = {
      storage: new Storage(),
      addingEvent: false,
      addingEventName: '',
      selectedDay: time,
      allItems: {},
      visItems: {},
    };
  }

    componentDidMount() {
      let time = this.state.selectedDay;

      this.state.storage._retrieveData('allItems').then(value => {
        const newItems = {};
        newItems[time] = [];
        if (value !== undefined) {
          this.setState({allItems: value});
          if (value[time]) {
            newItems[time] = value[time].items;
          }
        }
        this.setState({visItems: newItems});
      })
    }

  render() {
    return (
      <View>
      { !this.state.addingEvent &&
        <View style={styles.container}>
          <Agenda firstDay={1} showWeekNumbers={true}
            items={this.state.visItems}
            markedDates={this.state.allItems}
            selected={this.state.selectedDay}
            onDayPress={this.onDayPress.bind(this)}
            renderItem={this.renderItem.bind(this)}
            renderEmptyDate={this.renderEmptyDate.bind(this)}
            renderEmptyData = {this.renderEmptyData.bind(this)}
            rowHasChanged={(r1, r2) => {return r1.text !== r2.text}} />
          <Button title='Add Event' onPress={() => {this.setState({addingEvent: true})}} />
        </View>
      }
      { this.state.addingEvent &&
        <View style={[styles.container, { justifyContent: 'center'}]}>
          <TextInput
            style={{height: 40}}
            onChangeText={text => this.setState({addingEventName: text})}
            placeholder='Event name' />
          <Button title='Add' onPress={this.addEvent.bind(this)} />
          <Button title='Cancel' onPress={() => {this.setState({addingEvent: false})}} />
        </View>
      }
      </View>
    );
  }

  onDayPress(day) {
    const time = this.timeToString(day.timestamp);
    const newItems = {};
    newItems[time] = [];
    if (this.state.allItems[time]) {
      newItems[time] = this.state.allItems[time].items;
    }
    this.setState({visItems: newItems, selectedDay: time});
  }
  timeToString(time) {
    const date = new Date(time);
    return date.toISOString().split('T')[0];
  }
  renderItem(item) {
    return (
      <View style={[styles.event]}>
        <Text>{item.text}</Text>
        {item.goal &&
          <Text>Working towards: {item.goal}</Text>
        }
        {item.contacts &&
          <View>
            <Text>With:</Text>
            <FlatList
              data={item.contacts}
              renderItem={({item}) => <Text>{item}</Text>}
              keyExtractor={(item, index) => index.toString()}>
            </FlatList>
          </View>
        }
      </View>
    )
  }
  renderEmptyDate() {
    return (
      <View><Text>You have no events today.</Text></View>
    )
  }
  renderEmptyData() {
    return (
      <View><Text>You have no events.</Text></View>
    )
  }
  addEvent() {
    let today = this.state.selectedDay;
    let items = [];
    if (this.state.allItems[today]) {
      items = this.state.allItems[today].items
    }

    let newItem = {text: this.state.addingEventName};

    let newAllItems = {};
    newAllItems[today] = {marked: true, items: [...items, newItem]};

    let newVisItems = {};
    newVisItems[today] = [...items, newItem];

    let itemsToSave = {...this.state.allItems, ...newAllItems};
    this.setState({addingEvent: false,
      allItems: itemsToSave,
      visItems: newVisItems,
    });

    this.state.storage._storeData('allItems', itemsToSave);
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 30,
    height: '100%',
  },
  event: {
    minHeight: 60,
    margin: 4,
    marginBottom: 0,
    backgroundColor: '#ddd',
  },
});
