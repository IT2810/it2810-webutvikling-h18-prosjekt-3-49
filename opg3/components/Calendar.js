import React, {Component} from 'react';
import {StyleSheet, Text, View, FlatList, Button, TextInput, Picker} from 'react-native';
import {Agenda} from 'react-native-calendars';
import Storage from './Storage.js'

export default class Calendar extends Component {
    constructor(props) {
        super(props);

        let time = new Date();
        time = time.toISOString().split('T')[0];

        this.state = {
            storage: new Storage(),
            addingEvent: false,
            addingEventName: '',
            selectedDay: time,
            allItems: {},
            visItems: {},
            contacts: {},
            goals: {},
        };
    }

    componentDidMount() {
        let time = this.state.selectedDay;

        this.state.storage._retrieveData('allItems').then(value => {
            const newItems = {};
            newItems[time] = [];
            if (value !== undefined) {
                this.setState({allItems: value});
                if (value[time]) {
                    newItems[time] = value[time].items;
                }
            }
            this.setState({visItems: newItems});
        });
        this.state.storage._retrieveData('contacts').then(value => {
            if (value !== undefined) {
                this.setState({
                    contacts: value,
                    selectedContact: value[0]
                });
            }
        });
        this.state.storage._retrieveData('goals').then(value => {
            if (value !== undefined) {
                this.setState({
                    goals: value,
                    selectedGoal: value[0]
                });
            }
        })
    }

    render() {
        return (
            <View>
                {!this.state.addingEvent &&
                <View style={styles.container}>
                    <Agenda firstDay={1} showWeekNumbers={true}
                            items={this.state.visItems}
                            markedDates={this.state.allItems}
                            selected={this.state.selectedDay}
                            onDayPress={this.onDayPress.bind(this)}
                            renderItem={this.renderItem.bind(this)}
                            renderEmptyDate={this.renderEmptyDate.bind(this)}
                            renderEmptyData={this.renderEmptyData.bind(this)}
                            rowHasChanged={(r1, r2) => {
                                return r1.text !== r2.text
                            }}/>
                    <Button title='Add Event' onPress={() => {
                        this.setState({addingEvent: true})
                    }}/>
                </View>
                }
                {this.state.addingEvent &&
                <View style={[styles.container, styles.centered]}>
                    <TextInput
                        style={{height: 40, borderColor: 'gray', borderWidth: 1}}
                        onChangeText={text => this.setState({addingEventName: text})}
                        placeholder='Event name'/>
                    {this.state.contacts && this.state.contacts.length > 0 &&
                    <Picker
                        selectedValue={this.state.selectedContact}
                        onValueChange={(itemValue, itemIndex) => this.setState({selectedContact: itemValue})}>
                        {this.state.contacts.map(contact => {
                            return (<Picker.Item key={Math.random()} label={contact.fname + " " + contact.lname}
                                                 value={contact.fname + " " + contact.lname}/>)
                        })}
                    </Picker>
                    }
                    {this.state.goals && this.state.goals.length > 0 &&
                    <Picker
                        selectedValue={this.state.selectedGoal}
                        onValueChange={(itemValue, itemIndex) => this.setState({selectedGoal: itemValue})}>
                        {this.state.goals.map(goal => {
                            return (<Picker.Item key={Math.random()} label={goal.tag}
                                                 value={goal.tag}/>)
                        })}
                    </Picker>
                    }
                    <Button color='#4a4' title='Add' onPress={this.addEvent.bind(this)}/>
                    <Button color='#c55' title='Cancel' onPress={() => {
                        this.setState({addingEvent: false})
                    }}/>
                </View>
                }
            </View>
        );
    }

    onDayPress(day) {
        const time = this.timeToString(day.timestamp);
        const newItems = {};
        newItems[time] = [];
        if (this.state.allItems[time]) {
            newItems[time] = this.state.allItems[time].items;
        }
        this.setState({visItems: newItems, selectedDay: time});
    }

    timeToString(time) {
        const date = new Date(time);
        return date.toISOString().split('T')[0];
    }

    renderItem(item) {
        return (
            <View style={[styles.event]}>
                <Text>{item.text}</Text>
                {item.goal &&
                <Text>Working towards: {item.goal.tag}</Text>
                }
                {item.contacts &&
                <View>
                    <Text>With: {item.contacts.fname} {item.contacts.lname}</Text>
                </View>
                }
            </View>
        )
    }

    renderEmptyDate() {
        return (
            <View><Text>You have no events today.</Text></View>
        )
    }

    renderEmptyData() {
        return (
            <View><Text>You have no events.</Text></View>
        )
    }

    addEvent() {
        let today = this.state.selectedDay;
        let items = [];
        if (this.state.allItems[today]) {
            items = this.state.allItems[today].items
        }
        let newItem = {
            text: this.state.addingEventName,
            contacts: this.state.selectedContact,
            goal: this.state.selectedGoal,
        };

        let newAllItems = {};
        newAllItems[today] = {marked: true, items: [...items, newItem]};

        let newVisItems = {};
        newVisItems[today] = [...items, newItem];

        let itemsToSave = {...this.state.allItems, ...newAllItems};
        this.setState({
            addingEvent: false,
            allItems: itemsToSave,
            visItems: newVisItems,
        });

        this.state.storage._storeData('allItems', itemsToSave);
    }
}

const styles = StyleSheet.create({
    container: {
        paddingTop: 30,
        height: '100%',
    },
    event: {
        minHeight: 80,
        margin: 4,
        marginBottom: 0,
        backgroundColor: '#ddd',
    },
    centered: {
        width: '70%',
        justifyContent: 'center',
        alignSelf: 'center',
    }
});
