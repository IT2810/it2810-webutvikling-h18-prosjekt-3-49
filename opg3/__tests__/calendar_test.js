import React from 'react';
import Calendar from "../Calendar";

import renderer from 'react-test-renderer';

test('renders correctly', () => {
    const tree = renderer.create(<Calendar />).toJSON();
    expect(tree).toMatchSnapshot();
});