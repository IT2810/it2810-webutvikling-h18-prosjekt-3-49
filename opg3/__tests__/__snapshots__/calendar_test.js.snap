// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "height": "100%",
        "paddingTop": 30,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          undefined,
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#f4f4f4",
            "flex": 1,
            "marginTop": 104,
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            You have no events.
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "bottom": 1230,
            "height": 0,
            "justifyContent": "flex-end",
            "overflow": "hidden",
            "position": "absolute",
            "transform": Array [
              Object {
                "translateY": 1230,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "bottom": 0,
              "flex": 1,
              "height": 24,
              "left": 0,
              "position": "absolute",
              "right": 0,
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "backgroundColor": "#f2F4f5",
                  "borderRadius": 3,
                  "height": 7,
                  "marginTop": 10,
                  "width": 38,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "flexDirection": "row",
            "height": 104,
            "justifyContent": "space-around",
            "left": 0,
            "marginLeft": 15,
            "marginRight": 15,
            "opacity": 0,
            "paddingBottom": 7,
            "paddingTop": 15,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateY": -104,
              },
            ],
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Mon
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Tue
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Wed
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Thu
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Fri
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Sat
        </Text>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#b6c1cd",
              "fontSize": 13,
              "textAlign": "center",
              "width": 32,
            }
          }
        >
          Sun
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add Event
        </Text>
      </View>
    </View>
  </View>
</View>
`;
